### PR TITLE
[PM-10900] Fix master password policy bypassed when org user has 2FA

### PR DESCRIPTION
**Pull request by LRNcardozoWDF**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-10900

## 📔 Objective

Fix the flow so the update master password screen is presented to the user. Added the `MasterPasswordPolicyResponseModel` to the `IdentityTokenRequestError` so we can have the master password policies the org has active in order to check if user's master password complies with them.
In case it doesn't comply, the update master password screen will be presented to the user.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/AuthService.swift:210]: ⛏️ Add docs and could we maintain consistency with other places where we are using `isPreAuth` instead of `preAuth`?

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/AuthService.swift:595]: 🎨 🤔 We're currently handling `IdentityTokenRequestError` inside `getIdentityTokenResponse` and even more we're also handling `.twoFactorRequired(...)` in there as well so in order to maintain that handling as centralized as possible I think this should be moved there. If the issue was that this is only required for login with master password, then just do an if check that the `authenticationMethod` is `.password` and then you can handle your specific case all in the same place.

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/AuthService.swift:624]: ⛏️ Maintain the same naming as we use elsewhere => `isPreAuth`

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/AuthService.swift:648]: 🤔 Given this is cached in the service and this is a Singleton, could it happen that the user enters with master password then this cache is set and in the 2FA view the user goes back and changes account and when logging in it applies this reset reason to a different account? In the StateService this is already taken care of because it's tied to the `userId`.

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/AuthService.swift:730]: 🤔 If the user chooses not to login with the current account, when is `forcePasswordResetReason` cleared?

**Review by fedemkr — CHANGES_REQUESTED** [review]

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/AuthService.swift:625]: 🎨 When multiple parameters are passed to a function we're rarely remove all parameter labels, so I would remove the `_` to be consistent with other places. Additionally, order parameters alphabetically to be consistent with other places as well.

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/AuthService.swift:637]: 🎨 One thing to improve this a bit is to have an `init?(responseModel: MasterPasswordPolicyResponseModel)` in `MasterPasswordPolicyOptions+Extensions` which would take care of building the options or returning `nil`.
```swift
init?(responseModel: MasterPasswordPolicyResponseModel) {
    guard let responseModel else { return nil }
    self.init(
        minComplexity: responseModel.minComplexity ?? 0,
        minLength: responseModel.minLength ?? 0,
        requireUpper: responseModel.requireUpper ?? false,
        requireLower: responseModel.requireLower ?? false,
        requireNumbers: responseModel.requireNumbers ?? false,
        requireSpecial: responseModel.requireSpecial ?? false,
        enforceOnLogin: responseModel.enforceOnLogin ?? false
    )
}
```
So then you can simplify this to:
```swift
let policy = MasterPasswordPolicyOptions(responseModel: masterPasswordPolicy)
```

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/AuthService.swift:910]: ⚠️ We try never to use `!` as it can be dangerous if, for some reason `masterPassword` doesn't have a value, so add an `if let` check or add a `guard let` check which logs the error if that happens depending on whether you consider an error should be logged if `masterPassword` is `nil` and provide some description on such error so if we encounter that in Crashlytics we understand the context.

**Review by fedemkr — CHANGES_REQUESTED** [review]
Looks better, found some things to improve. Also would love if @matt-livefront and/or @KatherineInCode reviews this in case I've missed something here given that it affects the login flow.

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (KatherineInCode) [reply BitwardenShared/Core/Auth/Services/AuthService.swift:625]: It's worth noting that on the whole the `_` convention was somewhat intended as a replacement for the old Objective-C `[object with:thing and:otherThing]` idea. Early Swift as I recall even turned this into `object(with:thing and:otherThing)`, which evolved into the `object(thing, and: otherThing)` that's more common today.

File: BitwardenShared/Core/Auth/Models/Response/IdentityTokenErrorModel.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Models/Response/IdentityTokenErrorModel.swift:52]: ⛏️ In this case is whether it's enabled or not:
```suggestion
    /// Whether authenticator provider is enabled.
    let authenticator: Bool
```

File: BitwardenShared/Core/Auth/Services/API/Auth/Requests/IdentityTokenRequest.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/API/Auth/Requests/IdentityTokenRequest.swift:83]: 🤔 I think we should also be checking `let errorModel.twoFactorProvidersData.providersAvailable` to actually check there are 2FA providers available in the response data. Because if the response sends an empty 2FA providers data, then the current code won't fail and it should.

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/AuthService.swift:906]: ⛏️ No need for this comment as it's obvious.
```suggestion
```

**Review by fedemkr — COMMENTED** [review]
Looks good, just some small things to adjust.
Also, check the Files annotations given that there are some from Codecov worth noting regarding missing tests; although I'll leave that up to you to which ones are important to add.

File: BitwardenShared/Core/Auth/Services/AuthService.swift
Comment (fedemkr) [inline BitwardenShared/Core/Auth/Services/AuthService.swift:936]: ⛏️ 
```suggestion
                    message: "Two Factor Providers data is nil or empty."
```

File: BitwardenShared/UI/Auth/Login/LoginProcessor.swift
Comment (fedemkr) [inline BitwardenShared/UI/Auth/Login/LoginProcessor.swift:163]: ⚠️ Either you handle the error (and it's logging here) or in the service. Now it's going to log the error twice, once in the service and once here. If you want just remove the check in the service and leave this one since you have a specific error case it's going to be easy to pin point the origin of the error.

File: BitwardenShared/UI/Auth/Login/LoginWithDevice/LoginWithDeviceProcessor.swift
Comment (fedemkr) [inline BitwardenShared/UI/Auth/Login/LoginWithDevice/LoginWithDeviceProcessor.swift:195]: ⚠️ Either you handle the error (and it's logging here) or in the service. Now it's going to log the error twice, once in the service and once here. If you want just remove the check in the service and leave this one since you have a specific error case it's going to be easy to pin point the origin of the error.

File: BitwardenShared/UI/Auth/Login/LoginProcessor.swift
Comment (LRNcardozoWDF) [reply BitwardenShared/UI/Auth/Login/LoginProcessor.swift:163]: I chose to handle the error here and remove the check in the service.

File: BitwardenShared/UI/Auth/Login/LoginWithDevice/LoginWithDeviceProcessor.swift
Comment (LRNcardozoWDF) [reply BitwardenShared/UI/Auth/Login/LoginWithDevice/LoginWithDeviceProcessor.swift:195]: I chose to handle the error here and remove the check in the service.

**Review by fedemkr — APPROVED** [review]
LGTM, but just in case wait for a second approval from @matt-livefront or @KatherineInCode in case I've missed something given that this affects directly to the login flow.

**Review by matt-livefront — APPROVED** [review]

File: BitwardenShared/Core/Auth/Models/Response/MasterPasswordPolicyResponseModel.swift
Comment (fedemkr) [reply BitwardenShared/Core/Auth/Models/Response/MasterPasswordPolicyResponseModel.swift]: @andrebispo5 I believe this is because `IdentityTokenErrorModel` doesn't conform to `JSONResponse`, thus it doesn't use the `pascalOrSnakeCaseDecoder`.
I think we could just change: 
`try? JSONDecoder().decode` 
for 
`try? JSONDecoder.pascalOrSnakeCaseDecoder.decode`
in here and that should fix the issue without the `CodingKeys`
https://github.com/bitwarden/ios/blob/1f97eabc83c4c8d61109036bf34af574d32f3c4e/BitwardenShared/Core/Auth/Services/API/Auth/Requests/IdentityTokenRequest.swift#L80-L84

Comment (fedemkr) [reply BitwardenShared/Core/Auth/Models/Response/MasterPasswordPolicyResponseModel.swift]: do you agree @matt-livefront ?

**Pull request merged**
